pkg/scheduler/cron.go

```diff
@@ -14,45 +14,129 @@
 package scheduler
 
 import (
+	"context"
 	"time"
 
+	"github.com/go-logr/zapr"
+	"github.com/pingcap/log"
+	perr "github.com/pkg/errors"
 	cron "github.com/robfig/cron/v3"
+	"go.uber.org/zap"
 
 	"github.com/chaos-mesh/chaosd/pkg/core"
 )
 
 type Scheduler struct {
 	*cron.Cron
+	expStore    core.ExperimentStore
+	expRunStore core.ExperimentRunStore
+	cronStore   CronStore
 }
 
 type CronJob struct {
-	experiment core.Experiment
-	run        func()
+	scheduler  *Scheduler
+	experiment *core.Experiment
+	run        func() error
+}
+
+func hasCronDurationExceeded(startedAt time.Time, duration time.Duration) bool {
+	return time.Until(startedAt.Add(duration)).Milliseconds() < 0
 }
 
-// TODO: on running, create new experiment run
 // TODO: write tests for it
-func (cj CronJob) Run() {
-	cj.run()
+func (cj *CronJob) Run() {
+	var newRun *core.ExperimentRun
+	defer func() {
+		var updErr error
+		if panicErr, ok := recover().(error); panicErr != nil && ok {
+			log.Error("scheduled run errored", zap.String("expId", cj.experiment.Uid), zap.Error(panicErr))
+			if newRun != nil {
+				updErr = cj.scheduler.expRunStore.Update(context.Background(), newRun.UID, core.RunFailed, panicErr.Error())
+			} else {
+				// cannot even create a new run, maybe due to config error
+				// so better to set ERROR on the experiment and remove from scheduler
+				_ = cj.scheduler.Remove(cj.experiment.ID)
+				updErr = cj.scheduler.expStore.Update(context.Background(), cj.experiment.Uid, core.Error, "", cj.experiment.RecoverCommand)
+			}
+		} else {
+			log.Info("scheduled run success", zap.String("expId", cj.experiment.Uid))
+			if newRun != nil {
+				updErr = cj.scheduler.expRunStore.Update(context.Background(), newRun.UID, core.RunSuccess, "")
+			} else {
+				log.Info("cron run finished")
+			}
+		}
+		if updErr != nil {
+			log.Error("failed to update in DB", zap.Error(updErr))
+		}
+	}()
+
+	log.Info("Started new run", zap.String("expId", cj.experiment.Uid))
+	cfg, err := cj.experiment.GetRequestCommand()
+	if err != nil {
+		panic(perr.WithStack(err))
+	}
+	cronDuration, err := cfg.ScheduleDuration()
+	if err != nil {
+		panic(perr.WithStack(err))
+	}
+	if cronDuration != nil && hasCronDurationExceeded(cj.experiment.CreatedAt, *cronDuration) {
```

Comment:
seems that the experiment can only run one time. For example, the `schedule` is `*/1 * * * *`, and the `duration` is `30s`, the experiment will run at 17:01:00, 17:02:00 ..., and will keep 30s every time, and then recover it.

but the `hasCronDurationExceeded` here will return true when last 30s, and then remove it.

Reply:
@WangXiangUSTC I tested it but it seems to run fine for me :thinking: 
```
[2021/05/07 13:45:50.671 +05:30] [INFO] [zapr.go:69] [added] [now=2021/05/07 08:15:50.671 +00:00] [entry=1] [next=2021/05/07 08:16:00.000 +00:00]
[GIN] 2021/05/07 - 13:45:50 | 200 |  161.479713ms |       127.0.0.1 | POST     "/api/attack/stress"
[2021/05/07 13:46:00.002 +05:30] [INFO] [zapr.go:69] [wake] [now=2021/05/07 08:16:00.001 +00:00]
[2021/05/07 13:46:00.002 +05:30] [INFO] [zapr.go:69] [run] [now=2021/05/07 08:16:00.001 +00:00] [entry=1] [next=2021/05/07 08:17:00.000 +00:00]
[2021/05/07 13:46:00.002 +05:30] [INFO] [cron.go:74] ["Started new run"] [expId=54d519bf-5458-4f2e-b811-ab0d8a1e71a6]
[2021/05/07 13:46:00.013 +05:30] [INFO] [cron.go:98] ["executing attack on new exp run"] [expRunUID=f7ea65e5-e506-4182-afd7-bae5f89ee4b1]
[2021/05/07 13:46:00.013 +05:30] [INFO] [stress.go:65] ["stressors normalize"] [arguments=" --vm 2 --vm-keep --vm-bytes 100000000"]
[2021/05/07 13:47:00.000 +05:30] [INFO] [zapr.go:69] [wake] [now=2021/05/07 08:17:00.000 +00:00]
[2021/05/07 13:47:00.000 +05:30] [INFO] [zapr.go:69] [run] [now=2021/05/07 08:17:00.000 +00:00] [entry=1] [next=2021/05/07 08:18:00.000 +00:00]
[2021/05/07 13:47:00.001 +05:30] [INFO] [cron.go:74] ["Started new run"] [expId=54d519bf-5458-4f2e-b811-ab0d8a1e71a6]
[2021/05/07 13:47:00.001 +05:30] [INFO] [zapr.go:69] [removed] [entry=1]
[2021/05/07 13:47:00.010 +05:30] [INFO] [cron.go:62] ["scheduled run success"] [expId=54d519bf-5458-4f2e-b811-ab0d8a1e71a6]
[2021/05/07 13:47:00.010 +05:30] [INFO] [cron.go:66] ["cron run finished"]
```

^ as seen in logs, it ran only once since the first run was only 10s after scheduling it.